clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1087,23 +1076,38 @@ private Record nextFetchedRecord() {
         }
 
         private List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
+            // Error when fetching the next record before deserialization.
+            if (cachedRecordFetchException != null && lastRecord == null)
+                throw cachedRecordFetchException;
+
             if (isFetched)
                 return Collections.emptyList();
 
             List<ConsumerRecord<K, V>> records = new ArrayList<>();
             try {
                 for (int i = 0; i < maxRecords; i++) {
-                    Record record = nextFetchedRecord();
+                    Record record = lastRecord;
+                    // Only move to next record if there was no exception in the last fetch. Otherwise we should
+                    // use the last record to do deserialization again.
+                    if (cachedRecordFetchException == null) {
+                        lastRecord = null;
+                        record = nextFetchedRecord();
```

Comment:
Intuitively, I would expect `cachedRecordFetchException` to be set to null on the next line.